hr_timesheet_purchase_order/models/hr_timesheet_recurrence.py

```diff
@@ -0,0 +1,362 @@
+            ("second", "Second"),
+            ("third", "Third"),
+            ("last", "Last"),
+        ]
+    )
+    repeat_weekday = fields.Selection(
+        [
+            ("mon", "Monday"),
+            ("tue", "Tuesday"),
+            ("wed", "Wednesday"),
+            ("thu", "Thursday"),
+            ("fri", "Friday"),
+            ("sat", "Saturday"),
+            ("sun", "Sunday"),
+        ],
+        string="Day Of The Week",
+        readonly=False,
+    )
+    repeat_month = fields.Selection(
+        [
+            ("january", "January"),
+            ("february", "February"),
+            ("march", "March"),
+            ("april", "April"),
+            ("may", "May"),
+            ("june", "June"),
+            ("july", "July"),
+            ("august", "August"),
+            ("september", "September"),
+            ("october", "October"),
+            ("november", "November"),
+            ("december", "December"),
+        ]
+    )
+
+    @api.constrains("repeat_unit", "mon", "tue", "wed", "thu", "fri", "sat", "sun")
+    def _check_recurrence_days(self):
+        for timesheet in self.filtered(lambda p: p.repeat_unit == "week"):
+            if not any([getattr(timesheet, attr_name) for attr_name in DAYS]):
+                raise ValidationError(_("You should select a least one day"))
+
+    @api.constrains("repeat_interval")
+    def _check_repeat_interval(self):
+        if self.filtered(lambda t: t.repeat_interval <= 0):
+            raise ValidationError(_("The interval should be greater than 0"))
+
+    @api.constrains("repeat_number", "repeat_type")
+    def _check_repeat_number(self):
+        if self.filtered(lambda t: t.repeat_type == "after" and t.repeat_number <= 0):
+            raise ValidationError(_("Should repeat at least once"))
+
+    @api.constrains("repeat_type", "repeat_until")
+    def _check_repeat_until_date(self):
+        today = fields.Date.today()
+        if self.filtered(lambda t: t.repeat_type == "until" and t.repeat_until < today):
+            raise ValidationError(_("The end date should be in the future"))
+
+    @api.constrains(
+        "repeat_unit", "repeat_on_month", "repeat_day", "repeat_type", "repeat_until"
+    )
+    def _check_repeat_until_month(self):
+        if self.filtered(
+            lambda r: r.repeat_type == "until"
+            and r.repeat_unit == "month"
+            and r.repeat_until
+            and r.repeat_on_month == "date"
+            and int(r.repeat_day) > r.repeat_until.day
+            and monthrange(r.repeat_until.year, r.repeat_until.month)[1]
+            != r.repeat_until.day
+        ):
+            raise ValidationError(
+                _(
+                    "The end date should be after the day of "
+                    "the month or the last day of the month"
+                )
+            )
+
+    @api.constrains("repeat_day", "repeat_month")
+    def _check_repeat_day_or_month(self):
+        if 0 > self.repeat_day or self.repeat_day > 31:
```

Comment:
Not all months have 31 days. Please add condition for this.

Reply:
okay, I'll fix it